Update GraphQL Recommended vs Opt-In requirements

abernix opened the pull request:
Fixes https://github.com/open-telemetry/semantic-conventions/issues/2985

## Changes

Changes the firmness for `graphql.document` from `Recommended` to `Opt-In`.

## Context

The `graphql.document` is user-inputted, often contains sensitive information, is potentially unbounded in length and is also high-cardinality in the same way as the existing `graphql.operation.name` (which is already warned about).

Put another way, `graphql.document` is a general liability to have listed as `Recommended` without serious infrastructure considerations and needs.  This makes it a better candidate for being an `Opt-In` attribute.  In our customer adoption of OpenTelemetry and GraphQL, we've found GraphQL customers following this configuration/instruction purely based on its naming in this (still, experimental) specification and not understanding the implications (or just not thinking them through).

In most all cases, the lesser liability of `graphql.operation.name` (The actual operation name) is sufficient, as in many GraphQL deployments there is a link between the two which can be correlated out of band (e.g., client code-bases, operation manifests, etc.).  While the operation name isn't without its risk, but it's more likely to be dozens of *bytes* of a limited character set rather than dozens, hundreds or potentially thousands of *kilobytes*.  In that regard, it's reasonable that `graphql.operation.name` be left as `Recommended` for user experience (though the argument could easily be made that it should also be `Opt-In`).

## Merge requirement checklist

* [x] [CONTRIBUTING.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md) guidelines followed.
* [x] Change log entry added, according to the guidelines in [When to add a changelog entry](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md#when-to-add-a-changelog-entry).
  * If your PR does not need a change log, start the PR title with `[chore]`
* [x] Links to the prototypes or existing instrumentations (when adding or changing conventions)

abernix committed:
[graphql] Change graphql.document requirement level to opt-in

Fixes #2985

abernix committed:
change_type is breaking, not enhancement

Co-authored-by: Joao Grassi <5938087+joaopgrassi@users.noreply.github.com>

abernix commented on .chloggen/abernix_graphql-spec-requirements.yaml:
```diff
@@ -0,0 +1,26 @@
+# your pull request title with [chore] or use the "Skip Changelog" label.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: enhancement
```
Sounds right to me.  Fixed!

lmolkova commented on model/graphql/registry.yaml:
```diff
@@ -31,5 +31,5 @@ groups:
         type: string
         stability: development
-        note: The value may be sanitized to exclude sensitive information.
+        note: The value SHOULD be sanitized to exclude sensitive information.
```
I agree with the intention, but I don't think it's implementable in the current shape.

some suggestions (please review my English):


```suggestion
        note: If instrumentation can reliably identify and redact sensitive information is SHOULD do it.
```

or 

```suggestion
        note: Application developers should sanitize this attribute in telemetry processing pipeline to exclude sensitive information.
```

(we don't use normative language when talking about app developers - we don't have authority over them)

abernix committed:
Fix normative language in graphql.document note

Directs the SHOULD at instrumentation authors rather than application
developers, per reviewer feedback -- the spec has no authority over
app developers.

Ref: https://github.com/open-telemetry/semantic-conventions/pull/3118#discussion_r2620956314

abernix commented:
Is there anything else I need to do to help here?

joaopgrassi commented:
@abernix not sure what is wrong with the action, maybe you can try rebasing/squashing the commits to see if it gets happy again? It seems your PR is not allowed for maintainers to change it, so I can't push to it.

abernix commented:
@joaopgrassi Hmm, I rebased and pushed and it still is failing.

I don't think it's related to my branch at all — it looks like an issue with the workflow that would affect any fork
PR?

I think it's something like this: To figure out what files changed, `tj-actions/changed-files` needs two things: the PR branch commits, and the current tip of upstream main to compare against.  The workflow checks out the fork repo (mine, to get the PR branch), but the fork **doesn't** have the latest upstream main commits — those only exist in this upstream.  So when the action tries to find the merge base, it can't locate the upstream commit and fails.

Perhaps one fix would be adding `use_rest_api: true` to the `tj-actions/changed-files` step, which uses the GitHub API instead of local git history and sidesteps the problem?

abernix commented:
I don't meant to claim understand what it means, but I will note that the failing CI check is not a "Required" check.
